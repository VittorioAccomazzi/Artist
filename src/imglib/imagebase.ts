import { textChangeRangeIsUnchanged } from "typescript";


interface TypedArray {
    readonly length : number;
    [n: number]: number;
    buffer : ArrayBuffer;
}

interface TypedArrayConstructor<T extends TypedArray> {
    new (buffer: ArrayBuffer, byteOffset?: number, length?: number): T;
}

export type ImageUint8  = ImageBase<Uint8Array,Uint8ArrayConstructor>
export type ImageUint16 = ImageBase<Uint16Array,Uint16ArrayConstructor>
export type ImageFloat32= ImageBase<Float32Array,Float32ArrayConstructor>

export class ImageFactory {

    static Uint8( width : number, height : number ) : ImageUint8 {
        return new ImageBase<Uint8Array,Uint8ArrayConstructor>(Uint8Array.BYTES_PER_ELEMENT, Uint8Array, width, height)
    }
    static Uint16( width : number, height : number ) : ImageUint16{
        return (new ImageBase<Uint16Array,Uint16ArrayConstructor>(Uint16Array.BYTES_PER_ELEMENT, Uint16Array, width, height)) 
    }
    static Float32( width : number, height : number ): ImageFloat32{
        return new ImageBase<Float32Array,Float32ArrayConstructor>(Float32Array.BYTES_PER_ELEMENT, Float32Array, width, height)
    }
}


class ImageBase<T extends TypedArray, C extends TypedArrayConstructor<T>> {

    private buffer : T
    private pixels : T [] 

    constructor( elementSize : number, constructor : C, width : number, height : number ) {
        let nPixels= width * height;
        let bytes  = nPixels * elementSize
        let buffer= new ArrayBuffer(bytes);
        this.pixels= Array.from(Array(height), (e,i)=> new constructor (buffer, width*elementSize*i, width) ) 
        this.buffer= new constructor(buffer, 0, nPixels)
    }


    /**
     * get type of pixel stored.
     */
    get imageType(): string {
        return this.pixels[0].constructor.name.replace('Array','')
    }

    /**
     * returns the current list of pixels.
     */
    get imagePixels(): T {
        return this.buffer 
    }

    /**
     *  get the image buffer
     */
    get imageBuffer() : ArrayBuffer {
        return this.buffer.buffer
    }

    /**
     * get image height
     */
    get height() : number {
        return this.pixels.length
    }

    /**
     * return image height
     */
    get width() : number {
        return this.pixels[0].length
    }

    /**
     * get a pixel value.
     * @param x row of the pixel queried
     * @param y column of the pixel queried
     */
    get( x: number, y: number ) : number {
        if( !this.checkBoundary(x,y)) throw new Error (`Invalid image coordinates queried: image size ${this.width}x${this.height} pixel ${x},${y}`)
        return this.pixels[y][x]
    }

    /**
     * get a row of pixels
     * @param y row of pixels to select.
     */
    getRow(y:number) : T {
        if( y<0 || y>this.height )throw new Error (`Invalid y coordinate queried : image height ${this.height} y ${y}`)
        return this.pixels[y]
    }

    /**
     * 
     * @param x row pixel to set
     * @param y column pixel to se
     * @param value  value to set
     */
    set(x:number, y:number, value : number) : void {
        if( !this.checkBoundary(x,y)) throw new Error (`Invalid image coordinates set: image size ${this.width}x${this.height} pixel ${x},${y}`)
        this.pixels[y][x]= value
    }

    /**
     * check that the value passed are valid image coordinates.
     * @param x row
     * @param y column
     */
    private checkBoundary(x: number, y : number ) : boolean {
        return x>=0 && x < this.width && y>=0 && y<this.height
    }

}