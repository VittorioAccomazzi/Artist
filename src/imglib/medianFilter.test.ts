import {ImageFactory, ImageUint16} from './imagebase'
import Histogram from './histogram'
import MedianFilter from './medianFilter'
import { displayHisto, getCanvases, hash, dumpCanvas, toSeqCanvas } from './testutils'
import CanvasUtils from './canvasUtils'

test('shall implement median filter',async ()=>{
    let width = 16
    let height= 16
    let image = ImageFactory.Uint16(width, height)
    let pixels= image.imagePixels
    pixels.forEach((v,i)=>pixels[i]=i)
    let hist = new Histogram(image) 
    expect(hist.maxValue).toBe(1) // histogram flat with max value 1
    pixels[0]=10
    hist = new Histogram(image) 
    expect(hist.maxValue).toBe(2)
    expect(hist.histogramBins[10]).toBe(2)

    let med1 = MedianFilter.Run( hist, 1 )
    expect(med1.histogramBins[10]).toBe(1)

    let med2 = MedianFilter.Run( hist, 2 )
    expect(med2.histogramBins[10]).toBe(1)
})

test('shall smooth the histogram values',async ()=>{
    for await ( const [canvas,ctx,name] of getCanvases() ) {
        let [hImg] = CanvasUtils.toHSV(toSeqCanvas(canvas))
        let hImg16 = hImg.convertTo('Uint16') // will truncate the values
        let uHist = new Histogram(hImg16 as ImageUint16)
        
        let uImage = displayHisto(uHist)
        // await dumpCanvas(uImage, `hist-${name}-unfiltered`)
        let hashU = await hash(uImage)
        expect(hashU).toMatchSnapshot()

        let fHist = MedianFilter.Run(uHist,5)
        let fImage= displayHisto(fHist)
        // await dumpCanvas(fImage, `hist-${name}-filtered`)
        let hashF = await hash(fImage)
        expect(hashF).toMatchSnapshot()

    }
})