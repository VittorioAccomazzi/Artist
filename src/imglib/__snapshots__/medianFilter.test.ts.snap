// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shall smooth the histogram values 1`] = `"ffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffc0fffffffffffffffcffffffffffffffc0fffffffffffffff83fffffffffffffc0ffffffffffffff0013ffffffffffff003fffffffffff"`;

exports[`shall smooth the histogram values 2`] = `"ffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc1ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffc0fffffffffffffff03fffffffffffffc07fffffffffffff0007ffffffffffff803fffffffffff"`;

exports[`shall smooth the histogram values 3`] = `"7fffffffffffffffffffffffffffffff7fffffffffffffffffffffffffffffff7fffffffffffffffffffffffffffffff7fffffffffffffffffffffffffffffff7fffffffffffffffffffffffffffffff7fffffffffffffffffffffffffffffff7fffffffffffffffffffffffffffffff7fffffffffffffffffffffffffffffff7fffffffffffffffffffffffffffffff7fffffffffffffffffffffffffffffff7fffffffffffffffffffffffffffffff7fffffffffffffffffffffffffffffff7fffffffffffffffffffffffffffffff7fffffffffffffffffffffffffffffff7fffffffffffffffffffffffffffffff7fffffffffffffffffffffffffffffff7fffffffffffffffffffffffffffffff7fffffffffffffffffffffffffffffff7fffffffffffffffffffffffffffffff7fffffffffffffffffffffffffffffff1ffffffffffffffffffffffffffffffd1ffffffffffffffffffffffffffffffd1ffffffffffffffffffffffffffffffd1ffffffffffffffffffffffffffffffd1ffffffffffffffffffffffffffffffd1ffffffffffffffffffffffffffffffd1ffffffffffffffffffffffffffffffd1ffffffffffffffffffffffffffffffd1ffffffffffffffffffffffffffffffd1ffffffffffffffffffffffffffffffd1ffffffffffffffffffffffffffffff81ffffffffffffffffffffffffffffff01ffffffffffffffffffffffffffffff01ffffffffffffffffffffffffffffff01ffffffffffffffffffffffffffffff01ffffffffffffffffffffffffffffff01ffffffffffffffffffffffffffffff01ffffffffffffffffffffffffffffff01ffffffffffffffffffffffffffffff01ffffffffffffffffffffffffffffff01ffffffffffffffffffffffffffffff01ffffffffffffffffffffffffffffff01ffffffffffffffffffffffffffffff01ffffffffffffffffffffffffffffff01ffffffffffffffffffffffffffffff01ffffffffffffffffffffffffffffff01ffffffffffffffffffffffffffffff01ffffffffffffffffffffffffffffff01ffffffffffffffffffffffffffffff01ffffffffffffffffffffffffffffff01ffffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff001fffffffffffffffffffffffffffff001fffffffffffffffffffffffffffff001fffffffffffffffffffffffffffff001fffffffffffffffffffffffffffff001fffffffffffffffffffffffffffff001fffffffffffffffffffffffffffff001fffffffffffffffffffffffffffff001fffffffffffffffffffffffffffff001fffffffffffffffffffffffffffff001fffffffffffffffffffffffffffff001fffffffffffffffffffffffffffff001ffffffffffffffffffffffffff9ff001ffffffffffffffffffffffffff9ff001ffffffffffffffffffffffffff97f001ffffffffffffffffffffffffff97f001ffffffffffffffffffffffffff97e001ffffffffffffffffffffffffff97e001ffffffffffffffffffffffffff978001ffffffffffffffffffffffffff978001ffffffffffffffffffffffffff938001ffffffffffffffffffffffffff938001ffffffffffffffffffffffffff938001ffffffffffffffffffffffffff838001ffffffffffffffffffffffffff838001ffffffffffffffffffffffffff838001ffffffffffffffffffffffffff838001ffffffffffffffffffffffffff838001ffffffffffffffffffffffffff038001ffffffffffffffffffffffffff038001ffffffffffffffffffffffffff018001ffffffffffffffffffffffffff018001ffffffffffffffffffffffffff010001ffffffffffffffffffffffffff000001efffffffffffffffffffffffff000000efffffffffffffffffffffffff000000efffffffffffffffffffffffff000000efffffffffffffffffffffffff000000cfffffffffffffffffffffffff000000c7fffffffffffffffffffffffe000000c7fffffffffffffffffffffffe000000c7fffffffffffffffffffffffe000000c7fffffffffffffffffffffffc000000c7fffffffffffffffffffffffc000000c7fffffffffffffffffffffffc000000c7fffffffffffffffffffffff8000000c7fffffffffffffffffffffff8000000c7fffffffffffffffffffffff800000007fffffffffffffffffffffff000000007fffffffffffffffffffffff000000007ffffffffffffffffffffffc000000007ffffffffffffffffffffffc000000007ffffffffffffffffffffff0000000007ffffffffffffffffffffff0000000003fffffffffffffffffffffc0000000003fffffffffffffffffffff80000000003ffffffffffffffffffffe00000000001fffffffffffffffffff800000000"`;

exports[`shall smooth the histogram values 4`] = `"fffffffffffffffffffffffffffffffcfffffffffffffffffffffffffffffffcfffffffffffffffffffffffffffffffcfffffffffffffffffffffffffffffffcfffffffffffffffffffffffffffffffcfffffffffffffffffffffffffffffffcfffffffffffffffffffffffffffffffcfffffffffffffffffffffffffffffffcfffffffffffffffffffffffffffffffcfffffffffffffffffffffffffffffffcfffffffffffffffffffffffffffffffcfffffffffffffffffffffffffffffffc3ffffffffffffffffffffffffffffff83ffffffffffffffffffffffffffffff83ffffffffffffffffffffffffffffff83ffffffffffffffffffffffffffffff83ffffffffffffffffffffffffffffff83ffffffffffffffffffffffffffffff83ffffffffffffffffffffffffffffff83ffffffffffffffffffffffffffffff81ffffffffffffffffffffffffffffff81ffffffffffffffffffffffffffffff81ffffffffffffffffffffffffffffff81ffffffffffffffffffffffffffffff81ffffffffffffffffffffffffffffff81ffffffffffffffffffffffffffffff81ffffffffffffffffffffffffffffff81ffffffffffffffffffffffffffffff80ffffffffffffffffffffffffffffff80ffffffffffffffffffffffffffffff80ffffffffffffffffffffffffffffff80ffffffffffffffffffffffffffffff80ffffffffffffffffffffffffffffff80ffffffffffffffffffffffffffffff80ffffffffffffffffffffffffffffff80ffffffffffffffffffffffffffffff80ffffffffffffffffffffffffffffff80ffffffffffffffffffffffffffffff80ffffffffffffffffffffffffffffff00ffffffffffffffffffffffffffffff00ffffffffffffffffffffffffffffff00ffffffffffffffffffffffffffffff00ffffffffffffffffffffffffffffff007fffffffffffffffffffffffffffff007fffffffffffffffffffffffffffff007fffffffffffffffffffffffffffff007fffffffffffffffffffffffffffff007fffffffffffffffffffffffffffff007fffffffffffffffffffffffffffff007fffffffffffffffffffffffffffff007fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003fffffffffffffffffffffffffffff003ffffffffffffffffffffffffffffe003ffffffffffffffffffffffffffffe003ffffffffffffffffffffffffffffe003ffffffffffffffffffffffffffffe001ffffffffffffffffffffffffffffe001ffffffffffffffffffffffffffffe001ffffffffffffffffffffffffffffe001ffffffffffffffffffffffffffffe001ffffffffffffffffffffffffffffe001ffffffffffffffffffffffffffffe001ffffffffffffffffffffffffffffe001ffffffffffffffffffffffffffefc001ffffffffffffffffffffffffffefc001ffffffffffffffffffffffffffefc001ffffffffffffffffffffffffffefc001ffffffffffffffffffffffffffefc001ffffffffffffffffffffffffffcfc001ffffffffffffffffffffffffff87c001ffffffffffffffffffffffffff87c001ffffffffffffffffffffffffff81c001ffffffffffffffffffffffffff818001ffffffffffffffffffffffffff818001ffffffffffffffffffffffffff818001ffffffffffffffffffffffffff810001ffffffffffffffffffffffffff810000ffffffffffffffffffffffffff800000ffffffffffffffffffffffffff800000ffffffffffffffffffffffffff800000ffffffffffffffffffffffffff000000ffffffffffffffffffffffffff000000ffffffffffffffffffffffffff000000ffffffffffffffffffffffffff000000fffffffffffffffffffffffffe000000fffffffffffffffffffffffffe000000fffffffffffffffffffffffffe000000fffffffffffffffffffffffffc000000fffffffffffffffffffffffffc000000fffffffffffffffffffffffffc000000fffffffffffffffffffffffffc000000fffffffffffffffffffffffff8000000fffffffffffffffffffffffff8000000fffffffffffffffffffffffff8000000fffffffffffffffffffffffff800000007fffffffffffffffffffffff800000007fffffffffffffffffffffff000000007fffffffffffffffffffffff000000007ffffffffffffffffffffffe000000007ffffffffffffffffffffffc000000007ffffffffffffffffffffff8000000007ffffffffffffffffffffff8000000007ffffffffffffffffffffff0000000003fffffffffffffffffffffe0000000003fffffffffffffffffffffc0000000003fffffffffffffffffffff80000000003ffffffffffffffffffffe00000000001fffffffffffffffffffc00000000"`;
